nixos/auto-upgrade: introduce the concept of desync, for often-failing packages

Ekleog opened the pull request:
This adds the concept of a desync to the auto-upgrade service. A desync is basically a sub-nixpkgs, that is allowed to stay behind when auto-upgrading the system.

This has been especially useful for me for home-manager, which relatively often fails building on the stable channel. With a desync containing just home-manager and the few packages I want it to have available, I have stopped seeing any failure of the auto-upgrade service: instead, the home-manager part of my system is the only one that stays 2-3 auto-upgrades behind, until the module gets quite quickly fixed upstream.

This is useful for people who don't want to baby-sit their auto-upgrade mechanism, but still want to monitor failures, to avoid having a system very outdated.

I've been using this exact file for something like 2 months now, and its been working great for me! I have baby-sat it daily for this time, and it has been working fine without a single issue for a month with daily auto-upgrades — in this month, there have been a few times where esp. home-manager stopped building and then resumed building.

As I know this is a relatively controversial change, WDYT? I'll leave this PR open for a while, and if it gathers enough positive attention, I'll write up some release notes to get it ready to land :smile: 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ekleog committed:
nixos/auto-upgrade: introduce the concept of desync, for often-failing packages

Ekleog committed:
nixos/auto-upgrade: run nixfmt

Ekleog commented:
Note for the reviewer: the first commit contains the actual changes; the second commit is a huge lot of noise caused by the requirement to run `nixfmt` on everything

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,81 @@
+
+    # Either clean up the .old files, or recover from them
+
+    ${lib.concatMapStrings (file: ''
+      if [ "$build_succeeded" != "true" ]; then
+        mv ${file}.old ${file}
+      else
+        rm -f ${file}.old
+      fi
+    '') (map desyncFile (lib.attrNames cfg.desync))}
```
It'd be nice to factor out the `if` condition, such that it isn't repeated for each desync file.

E.g. using a bash function:
```sh
# Either clean up the .old files, or recover from them
if [ "$build_succeeded" != "true" ]; then
  cleanup() {
    mv "$1.old" "$1"
  }
else
  cleanup() {
    rm -f "$1.old"
  }
fi
```

```nix
lib.concatMapStringsSep "\n"
  (name: "cleanup ${desyncFile name}")
  (builtins.attrNames cfg.desync)
```

Ekleog replied:
That's an interesting idea! I didn't like the repetition either, but this seems like it should work :smile:

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+                      };
+                in
+                {
+                  desyncAge = desync.desyncAge;
+                  path = desync.path;
```
```suggestion
                  inherit (desync) desyncAge path;
```

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+                {
+                  desyncAge = desync.desyncAge;
+                  path = desync.path;
+                  pkgs = config.pkgsFor desync.path;
```
Should this be:

```suggestion
                  pkgs = config.pkgsFor config.path;
```

?

Ekleog replied:
The goal here is to have an evaluated version of the packages that is from the latest working desync version. This version is recorded in the desync file, hence this choice :)

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+                    set -euo pipefail
+
+                    ${lib.optionalString (config.desyncAge >= config.maxDesyncAge) ''
+                      echo "Did not get any successful build for ${name} in the past ${config.desyncAge} rebuilds!" >&2
```
```suggestion
                      echo "Did not get any successful build for ${name} in the past ${toString config.desyncAge} rebuilds!" >&2
```

`desyncAge` is an int, so it can't be interpolated into a string without first using `builtins.toString`.

Ekleog replied:
Good catch! You're right that I never actually hit this case in my testing… someday nix will be a compiled language :smiling_face_with_tear:

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -189,6 +494,8 @@ in
       }
     ];
 
+    environment.systemPackages = lib.optional (cfg.desync != { }) upgradeDesyncsPkg;
```
Probably better to use `mkIf` than `optional`. Unless this was for a specific reason such as avoiding inf-rec.

```suggestion
    environment.systemPackages = lib.mkIf (cfg.desync != { }) [
      upgradeDesyncsPkg
    ];
```

Ekleog replied:
I'm not sure I see the advantage of using `mkIf` here?

MattSturgeon replied:
`optional false` results in a definition of `[ ]` while `mkIf false` (usually\*) results in no definition.

Practically speaking, there's not _much_ difference to the final value. But there is a subtle difference in how the definitions get merged, how they'd appear in warnings, etc.

\* one exception being "lazy" types that have an `emptyValue` used when undefined

Ekleog replied:
That's actually why I prefer `optional` whenever possible: IIRC `mkIf false` results in something like `{ if = false; value = <thunk>; }`, which I'd expect would slow down eval. While otherwise there would be basically no practical impact.

Does that make sense, or has the module system changed too much since I last looked at it?

MattSturgeon replied:
You are correct, `mkIf` will wrap the value. However imo it's better to try to use the module system idiomatically, rather than attempt to micro-optimise these kinda things.

As for which is actually more optimal, I'm unsure. On the one hand `mkIf false` will result in one less definition being passed to the type's merge function. On the other hand, merging an empty list as one of the definitions shouldn't be expensive.

---

One scenario where `mkIf false` is better than an unconditional "empty" definition is when dealing with multiple override priorities.

E.g. a user defines `mkDefault [ "foo" ]` and some other module defines `mkIf false [ "bar" ]`. Due to using `mkIf`, the second definition is skipped and the `highestPrio` is `(mkDefault null).priority` (1000). Therefore the final value is `[ "foo" ]`.

If we tweak that example to use `if then else` or `optionals` instead of `mkIf`, then the second definition is now unconditional\* and will always override the `mkDefault` definition. Therefore the `highestPrio` will be `defaultOverridePriority` (100) and the final value will be `[ ]`.

\* the definition's _value_ is conditional, but the definition itself is not.

---

Although, in this specific scenario, it seems almost impossible that `environment.systemPackages` would ever end up with a lower priority than `defaultOverridePriority`, given it is defined by the majority of modules.

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,81 @@
+        exec ${upgradeDesyncsPkg}/bin/nixos-upgrade-desyncs
+      ''}
+    '';
+  desyncFile = c: "/var/lib/nixos/desyncs/${lib.strings.escapeShellArg c}.nix";
```
Why `c`? Can we name it something more descriptive?

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,81 @@
+    # Generate the new files
+
+    ${lib.concatMapStrings (
+      c:
```
Same

MattSturgeon reviewed (commented):
I'm not very familiar with auto-upgrade, but I have some minor feedback:

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+              options.overlays = lib.mkOption {
+                default = [ ];
+                description = "Overlays to add to the nixpkgs evaluation for this desync";
+                # TODO: when upstreaming, factor out with the `overlayType` in nixos/modules/misc/nixpkgs.nix by putting it in lib/types.nix
```
Thank you! I had completely forgotten about this `TODO`, that I couldn't do initially as I overrode the module locally :smile:

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+                  This defaults to `14` in order to avoid packages silently becoming infinitely old with
+                  this module simply enabled.
+                '';
+                type = lib.types.nullOr lib.types.int;
```
Interesting, I hadn't known about it! As for having `0` meaning "no limit", in my opinion it's bad software design — 0 should mean 0 when it's used, and the legacy of it meaning "unlimited" should vanish as early as C hopefully will.

Anyway, I migrated it to a positive integer, thank you!

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+                # Derivation called to upgrade the desync files.
+                # `$out` will be symlinked from `/var/lib/nixos/desyncs/${name}.nix by the `upgradeScript`
+                readOnly = true;
+                visible = false;
```
Interesting, I've never looked in details in what the semantics are here — applied with both flags in both cases, hopefully everything will work fine :)

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+              config =
+                let
+                  desync =
+                    if builtins.pathExists /var/lib/nixos/desyncs/${name}.nix then
+                      import /var/lib/nixos/desyncs/${name}.nix
```
It does, but auto-upgrade already assumes channels AFAICT — there do seem to be something with flakes, but TBH it wouldn't surprise me if it were not possible to have this feature with flakes anyway… I must say I gave up on following all the flakes story quite a few years ago :laughing:

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +252,233 @@ in
+                      desyncAge = 0;
+                      path = builtins.toPath ${
+                        # Convert the now-validated path into an absolute path in the store, then into a string
+                        lib.strings.escapeNixString (builtins.path { path = config.defaultPath; })
```
AFAICT here `config.defaultPath`, is actually a string, and not a path. Because that's the way the it's being defined by the module system. Plus it doesn't actually point into the store, but in /nix/var/something/channels directory, IIRC.

So I do need to turn it into a path, so that it gets added (to the store if required and) as a context to the string, I think :)

Ekleog committed:
handle review comments

Ekleog committed:
nixfmt

roberth reviewed (commented):
This looks very useful, thank you!

SuperSandro2000 commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,84 @@
+        # This is expected, as it evaluates the nixos configuration in exactly
+        # the same way as `nixos-rebuild` would have.
+        cp -f ${file} ${file}.old || true
+        ${pkgs.nix}/bin/nix-build \
+          -E '(import <nixpkgs/nixos> {})
+          .config.system.autoUpgrade.desync
+          .'${lib.strings.escapeShellArg (lib.strings.escapeNixString desyncId)}'.desyncUpgradeScript' \
+          --out-link ${file} \
+          || \
+        ${pkgs.nix}/bin/nix-build \
```
```suggestion
        ${lib.getExe' config.nix.package "nix-build"} \
          -E '(import <nixpkgs/nixos> {})
          .config.system.autoUpgrade.desync
          .'${lib.strings.escapeShellArg (lib.strings.escapeNixString desyncId)}'.desyncUpgradeScript' \
          --out-link ${file} \
          || \
        ${lib.getExe' config.nix.package "nix-build"} \
```

SuperSandro2000 commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,84 @@
+      ''}
+    '';
+  desyncFile = desyncId: "/var/lib/nixos/desyncs/${lib.strings.escapeShellArg desyncId}.nix";
+  upgradeDesyncsPkg = pkgs.writeScriptBin "nixos-upgrade-desyncs" ''
+    #!/bin/sh
+    set -euo pipefail
```
```suggestion
  upgradeDesyncsPkg = pkgs.writeShellScriptBin "nixos-upgrade-desyncs" ''
    set -euo pipefail
```

SuperSandro2000 commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,84 @@
+  upgradeFlag = if cfg.upgradeAll then "--upgrade-all" else "--upgrade";
+  upgradeScript =
+    flags:
+    pkgs.writeScript "upgrade-channels" ''
+      #!/bin/sh
```
```suggestion
    pkgs.writeShellScript "upgrade-channels" ''
```

SuperSandro2000 commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,84 @@
 let
   cfg = config.system.autoUpgrade;
+  nixpkgs-cfg = config.nixpkgs;
+  nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
```
```suggestion
  nixos-rebuild = lib.getExe config.system.build.nixos-rebuild;
```

SuperSandro2000 commented:
I will call it here: that will unexpectedly break your setup and the example alone is already enough: What if matrix-synapse has an update that changes the module in a backwards incompatible way? Then unexpected things happen.

roberth commented:
> I will call it here: that will unexpectedly break your setup and the example alone is already enough: What if matrix-synapse has an update that changes the module in a backwards incompatible way? Then unexpected things happen.

Can be mitigated with clear documentation and the inclusion of tests to gate the upgrade.
Only mitigated though.
OTOH for something like a browser or a CLI tool, it should be rare to have bad interactions and breakage.

SuperSandro2000 commented:
A browser is another bad example like most graphical programs. Especially QT stuff super easily breaks mixing versions. Similar things for hardware accelerated things.

MattSturgeon commented on lib/types.nix:
```diff
@@ -882,6 +882,13 @@ rec {
       nestedTypes.elemType = elemType;
     };
 
+    overlayType = lib.mkOptionType {
+      name = "nixpkgs-overlay";
+      description = "nixpkgs overlay";
+      check = lib.isFunction;
+      merge = lib.mergeOneOption;
+    };
+
```
> `mkOrder order`

Did you mean `mkOverride priority`?

roberth replied:
`mkOverride` replaces the entire value, so the type doesn't need to merge anything. I really meant `mkOrder`, which creates an explicit ordering. Found the issue
- https://github.com/NixOS/nixpkgs/issues/215486

```suggestion
    extensionFunction = lib.mkOptionType {
      name = "extensionFunction";
      description = "extension function";
      check = lib.isFunction;
      # A more capable function is possible, see
      # https://github.com/NixOS/nixpkgs/issues/215486
      merge = lib.mergeOneOption;
    };

```

MattSturgeon replied:
Maybe I'm misinterpreting, but wouldn't [`mergeOneOption`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.options.mergeOneOption) throw if there are multiple definitions of the same `mkOverride` priority, regardless of whether they have different `mkOrder` priorities?

It sounds like what we want here is a merge function that asserts each definition has a unique `mkOrder` priority? Not in scope for this PR, but should probably be reflected in the comment.

Ekleog replied:
Thank you! Fixed :)

roberth commented:
> A browser is another bad example like most graphical programs. Especially QT stuff super easily breaks mixing versions. Similar things for hardware accelerated things.

IIUC, these aren't true overlays that use the main Nixpkgs' dependencies; the packages bring their own closure with dependencies from the Nixpkgs version in which they were packaged. It'd have to be a runtime problem. I have a home-manager that's ahead of NixOS, and it seems to work fine.

Hardware acceleration is known to cause problems, but isn't that mostly a Nix-on-other-Linux problem? On NixOS I believe this is mostly solved with the `/run` symlinks.

Anyway, I might just be lucky or misinformed or something, and it's good to err on the side of caution. Feel free to use simpler examples, e.g. just "basic CLI tools"

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -1,12 +1,84 @@
 {
```
What you're mentioning is exactly what I tried to implement indeed! Just, I'm not sure I understand your point 4 :sweat_smile: 

In what you're suggesting, I guess `track` and `gate` are the two I'd prefer! I'd still love to hear feedback from other people, I hadn't noticed the negative connotation to `desync` at all :)

Ekleog commented:
> I will call it here: that will unexpectedly break your setup and the example alone is already enough: What if matrix-synapse has an update that changes the module in a backwards incompatible way? Then unexpected things happen.

It only breaks if you use the matrix-synapse from the desync in an overlay _and_ matrix-synapse stops building _and_ there is a backward-incompatible change _exactly_ at that time. A backward-incompatible change that I'd expect has probably never happened yet in the time desyncs will usually live, ie. 14 days by default. For a module to be incompatible with a package 14 days old, it needs to be very seriously unstable.

Also, if I had handled that issue manually, I would do exactly that: use a a-few-days-old version of matrix-synapse until the rest is fixed, so that I could get the security fixes that land on nixpkgs — the gain/risk ratio is clearly beneficial IMO.

If you really fear the module being incompatible with the desync's packages, then you can also disable the matrix-synapse module from the outer nixpkgs, and import it from the desync's path instead. TBH I personally don't see the point, but you can. Do you want to update the example to give that example? (Disclaimer: I did test on my servers with exactly matrix-synapse and home-assistant, a scheme with a disabledModule would be untested from my end)

Ekleog committed:
review comments

SuperSandro2000 commented:
> For a module to be incompatible with a package 14 days old, it needs to be very seriously unstable.

No, it doesn't. The smallest change in any module can already trigger serious bugs. Already a missing ``proxyWebsockets = true`` like which was required to get mobilizon back to work https://github.com/NixOS/nixpkgs/commit/611f4847de16748192a5103dacc15f2657c55fc6 can already break everything and render the program next to unusable.

> The only case I could think of would be a program completely changed its configuration, and the derivation has been updated alongside the module, to keep things working. This would happen only with pretty unstable packages, that I'd expect would also break in other ways relatively regularly.

The program doesn't need to completely change its config. It is enough to change one bit of its config which doesn't even need to be considered a breaking change by up stream. This could be as easy as a changed default which conflicts with NixOS, a new required nginx rule or a slightly changed regex for a location. Such changes happen to eg Nextcloud somewhat regular (https://github.com/NixOS/nixpkgs/commit/71dd9c3d538195abdcdaf6b1afdcf72193dcaf0a) or a changes like https://github.com/NixOS/nixpkgs/commit/c82195d9e82b28147e9a960f6d47ca9200228452  or https://github.com/NixOS/nixpkgs/commit/611f4847de16748192a5103dacc15f2657c55fc6 .

> Also, and more importantly, if I had handled that issue manually, I would do exactly what desyncs do: use a a-few-days-old version of matrix-synapse until the rest is fixed, so that I could get the security fixes that land on nixpkgs — the gain/risk ratio is clearly beneficial IMO.

But then you are on your own and everyone will tell you that this is generally a bad idea to do. If issues like I linked above occur because of this module, we are probably unable to debug them or to provide any kind of support as we are basically undermining the basic promise of NixOS: to have a declarative config with one definitive source. We cannot promise this then anymore. Basically we would need to reject any issue that occurs when this module is enabled.

The more I think about this, the more I am against this insanity.

Ekleog commented:
> we are basically undermining the basic promise of NixOS: to have a declarative config with one definitive source

I think this is the source of our disagreement: we interpret "one definitive source" differently. To you, this source seems to be nixpkgs. To me, this source is my configuration. I have very often had packages from old nixpkgs releases on my system (IIRC, in 2021 I had a package from nixpkgs release 17.something). And this is part of what I was promised when I started using nix & nixos, and what makes me keep using it.

The potential breakage you're mentioning are indeed possibilities. Things that could happen someday in the future, maybe.

On the other hand, systems that do not receive security updates for public vulnerabilities is a problem that we're seeing today. Vulnerabilities that probably have botnets scanning for them all over the internet.

If we can reduce the exposure time by even a few hours, it's a great win. Hydra already takes a long time to move forward, so if we need to wait for multiple evals just because home-assistant was broken at the wrong commit, it can easily leave systems vulnerable for a week, or even more. Which is enough for the system to be compromised.

This being said people won't configure a desync for every single of their packages. If you try out this PR, you can see it does take a bit of time and work to configure one properly, so you can feel assured that it won't be used except for repeat offenders (in my case, matrix-synapse and home-assistant, though I'm thinking of removing synapse as it seems to have stabilized better recently).

-----

This being said, I also want to point out that people could use desyncs to setup nixos declarative containers, with eg. `containers.<name>.nixpkgs = services.auto-upgrade.desync.matrix-synapse.pkgs` (pseudo-code, I'm on my phone).

It's not literally speaking what I tested on my system so I can't guarantee it'll work fine, but I can see no reason why it shouldn't.

Would updating the examples to suggest this option feel better to you? This way, modules and packages would be from the same checkout, using the made-for-that nixos declarative containers.

Ekleog commented:
> we are basically undermining the basic promise of NixOS: to have a declarative config with one definitive source

I think this is the source of our disagreement: we interpret "one definitive source" differently. To you, this source seems to be nixpkgs. To me, this source is my configuration. I have very often had packages from old nixpkgs releases on my system (IIRC, in 2021 I had a package from nixpkgs release 17.something). And this is part of what I was promised when I started using nix & nixos, and what makes me keep using it.

The potential breakage you're mentioning are indeed possibilities. Things that could happen someday in the future, maybe.

On the other hand, systems that do not receive security updates for public vulnerabilities is a problem that we're seeing today. Vulnerabilities that probably have botnets scanning for them all over the internet.

If we can reduce the exposure time by even a few hours, it's a great win. Hydra already takes a long time to move forward, so if we need to wait for multiple evals just because home-assistant was broken at the wrong commit, it can easily leave systems vulnerable for a week, or even more. Which is enough for the system to be compromised.

This being said people won't configure a desync for every single of their packages. If you try out this PR, you can see it does take a bit of time and work to configure one properly, so you can feel assured that it won't be used except for repeat offenders (in my case, matrix-synapse and home-assistant, though I'm thinking of removing synapse as it seems to have stabilized better recently).

-----

This being said, I also want to point out that people could use desyncs to setup nixos declarative containers, with eg. `containers.<name>.nixpkgs = services.auto-upgrade.desync.matrix-synapse.pkgs` (pseudo-code, I'm on my phone).

It's not literally speaking what I tested on my system so I can't guarantee it'll work fine, but I can see no reason why it shouldn't.

Would updating the examples to suggest this option feel better to you? This way, modules and packages would be from the same checkout, using the made-for-that nixos declarative containers.

Ekleog commented:
Unrelatedly: I'm considering adding a flag to add `--max-jons 0` to the nix-build call, so that it fails immediately if the substituter fails. This might help avoid my rpi taking 3-4 hours every night to attempt rebuilding hass when it's broken. The default would still be to attempt a local build first.

roberth commented:
> add `--max-jons 0` to the nix-build call

Freudian slip?

Ekleog committed:
use a declarative container for the example

Ekleog committed:
also document the dangers of mix-and-matching packages

Ekleog commented:
@SuperSandro2000 I just pushed a change to the examples, to only use declarative containers, as well as document the dangers of mix-and-matching packages. Does this look better to you? (Disclaimer: it's not what I use on my production systems, because moving things towards a declarative container takes more work than I'm currently willing to allocate to it. But I see no reason why it wouldn't work)

-------

For everyone, WDYT about the names? Please emoji:
- :+1: for `desync`
- :smile: for `track`
- :rocket: for `gate`

Ekleog committed:
fix typo

Ekleog commented:
Good news: I just moved my cups to a desync'd container to work around https://github.com/NixOS/nixpkgs/pull/391529 that's been failing for almost a week; and it seems to be working fine :tada: I just had to adjust a bit the example for it to actually work :smile:

Ekleog commented:
Additional idea: we could add an `autoDesync = true` to nixos declarative containers, that'd automatically desync with the container's start script, which should be enough to desync iff the container stops building.

Ekleog committed:
symlink instead of copying out of the store, to keep the gc root alive when a rollback is needed

Ekleog committed:
avoid remembering /nix/var/nix if a desync literally never built fine

Ekleog committed:
Merge remote-tracking branch 'upstream/master' into desync

Ekleog committed:
write release notes

Ekleog commented:
Ok, AFAICT this is ready to land. I just synced my locally-overridden module with this one, and adjusted the failure mode I had discovered after a while. I also wrote the pending release notes.

As future work I'll maybe add the `containers.<name>.autoDesync` option, but I won't be doing that as part of this PR just yet, to avoid landing too much at once.

AFAICT all negative comments have been handled, with the documentation now very explicit on the potential issues that could arise, and only recommending known-good setups. The only remaining question is the naming of the option, that [seems to have no clear winner](https://github.com/NixOS/nixpkgs/pull/385143#issuecomment-2726010299).

I also upped the PR in the matrix communities a few times with no new comments, so if I don't hear someone yelling at me within like a week, I'll probably land this :smile:

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +253,251 @@ in
+          If no such version has a passing build for one week, then the `mixos-rebuild`
+          service will still fail.
+
+          Finally, note that overlays are not taken into account inside desyncs. This is
+          because, otherwise, evaluation would enter infinite loops.
+          If you need to overlay your desyncs, you can use the `desync.*.overlays` option.
```
Is there a reason we can't just read the `nixpkgs.overlays` option instead of introducing a new `desync.*.overlays` option?

If the new option stays, this reference to it should ideally be a markdown link. Also, this again feels like maybe it should be an admonition.

MattSturgeon replied:
Oh, I see. You're using the `nixpkgs.overlays` to implement your desynced packages.

Maybe we could avoid this inf-rec by having an internal option: `nixpkgs.finalOverlays` or something?

The `nixpkgs` module would simply define `nixpkgs.finalOverlays = cfg.overlays`, while other internal implementation, like the desync module, could additionally write to the `nixpkgs.finalOverlays` option, while still reading from the `nixpkgs.overlays` option.

Even so, it feels like some users would want their overlays to be considered in desyncs while other users may not. So we'd probably also need some boolean toggle.

I guess this idea is fine to be left out of scope for this initial PR. If you like the idea, maybe add a TODO comment?

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +253,251 @@ in
+          If they build fine then they will be upgraded, if not they will be left at their
+          current setting.
+
+          Using `maxDesyncAge`, you can limit the number of rebuilds that one nixpkgs
```
If `maxDesyncAge` is an option, could you link to it?

MattSturgeon replied:
You'd need to use the relevant "anchor" tag from the [nixos manual's options appendix](https://nixos.org/manual/nixos/unstable/options).

It should be `#opt-system.autoUpgrade.desync.maxDesyncAge`, but to be sure you can build the manual locally using `nix-shell ./nixos/doc --run devmode`.

The syntax will look like ``[`maxDesyncAge`](#opt-system.autoUpgrade.desync.maxDesyncAge)`` or `[](#opt-system.autoUpgrade.desync.maxDesyncAge)` to automatically use the full option name as the link text.

Ekleog replied:
Oh yes now that you say it that does ring a bell, thank you! I'm surprised that there aren't more options making use of that, if there were enough examples to copy-paste from it'd make things much easier… well at least after this PR the auto-upgrade module will be such an example!

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -174,6 +253,251 @@ in
+                    set -euo pipefail
+
+                    # Force dependencies on `requireBuilds`
+                    # ${lib.concatStringsSep " " (config.requireBuilds (config.pkgsFor (config.defaultPath)))}
```
Up to you, but sometimes these deeply nested function applications can be easier to read as a pipe:

```suggestion
                    # ${lib.pipe config.defaultPath [
                      config.pkgsFor
                      config.requireBuilds
                      (lib.concatStringsSep " ")
                    ]}
```

Reading it again makes me question whether `pkgsFor` and `requireBuilds` are actually functions, though? Just based on their naming convention, I would assume `requireBuilds` is a boolean, and `pkgsFor` is either a map/dict attrset or a function of type `desync -> nixpkgs-instance`.

Ekleog replied:
One day... one day we'll have [a $ operator](https://github.com/NixOS/nix/issues/1845). But I don't have enough energy to start a new RFC now so 🤷 

As for the question about pkgsFor and requireBuilds, they must take the package set as a parameter. The rationale here is, that the desync system will try bumping the channel, run these two functions, if everything builds fine then accept the new version, and if not actually desync from the channel.

Technically speaking they're always used with `requireBuilds (pkgsFor path)`, but I have `pkgsFor` into a separate option so that it can have the nixpkgs default, which should be more convenient in most cases.

Does that make sense to you?

MattSturgeon replied:
> One day... one day we'll have [a $ operator](https://github.com/NixOS/nix/issues/1845).

We are at least [getting](https://github.com/NixOS/rfcs/pull/148) a `|>` pipe operator :tada:

Ekleog committed:
review comments

Ekleog committed:
Merge remote-tracking branch 'upstream/master' into desync

Ekleog committed:
fix manual build

Ekleog committed:
nixfmt

Ekleog committed:
remove example tag

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
````diff
@@ -174,6 +253,251 @@ in
+          because, otherwise, evaluation would enter infinite loops.
+          If you need to overlay your desyncs, you can use the `desync.*.overlays` option.
+
+          For example, you can use this module this way, in order to get your system to
+          auto-upgrade even if `matrix-synapse` does not build on the latest channel:
+          ```nix
+          {
+            system.autoUpgrade.desync.for-synapse = {
+              requireBuilds = pkgs: [ pkgs.matrix-synapse ];
+            };
+
+            containers.matrix-synapse = {
+              autoStart = true;
+              nixpkgs = config.system.autoUpgrade.desync.for-synapse.path;
+              config = {
+                services.matrix-synapse = {
+                  # ...
+                };
+              };
+            };
+          }
+          ```
````
Unfortunately, it looks like this leads to CI failure: this requires an "example title", and when I try to add one like the ones already there in the rest of the codebase I fall upon https://github.com/NixOS/nixpkgs/issues/214209 :smiling_face_with_tear:

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -184,6 +265,255 @@ in
+                      cat - > "$out" <<EOF
+                    {
+                      desyncAge = ${toString (config.desyncAge + 1)};
+                      path = builtins.toPath ${lib.strings.escapeNixString config.path};
```
Could something like this be where the trailing `/` error is coming from?

MattSturgeon commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -184,6 +265,255 @@ in
+                      desyncAge = 0;
+                      path = builtins.toPath ${
+                        # Convert the now-validated path into an absolute path in the store, then into a string
+                        lib.strings.escapeNixString (builtins.path { path = config.defaultPath; })
```
Or here, maybe?

ElvishJerricco reviewed (dismissed):
Seeing this PR for the first time now.

---

The commit history in this PR needs some work.

- You should rebase your branch on `upstream/master`, not merge it.
- The files are not formatted, neither in the current state (which itself is a blocker), nor in the individual commits. Nixpkgs is 100% formatted now. So should be every one of your commits. Thus, commits called `nixfmt` should be unnecessary.
- Commits called "[review comments](https://github.com/NixOS/nixpkgs/pull/385143/commits/463f7689c6605db63434ac1b338c9d7430275d0c)" should not exist. Such changes should be squashed into the relevant original commit.

---

Perhaps more important than any of my other review comments is this: I do not think this concept should be merged at all. Not in this form nor in any form with my other comments addressed. I'm sorry for the blunt and negative tone. I don't mean to say your contribution is a poor one. I just don't think it's a good fit for NixOS.

This is a serious departure from NixOS's ordinary functioning. The evaluation of a NixOS config is generally considered to be stateless, with the primary exception being the channels managed directly by the user / `nix-channel`. A NixOS config can generally be evaluated and even built on separate machines before being deployed, which cannot work with "desyncs". The near-statelessness of NixOS is one of its key selling points.

I also question the utility of this. The first time a package breaks, you have to rollback your channel, add the desync, and then attempt the upgrade again. In other words, when an unexpected breakage occurs, you still have to perform manual intervention. If the point is that there are packages that break so often that you'd like to have a desync ready for them perpetually, I think that's more of a sign that the package is not well-maintained and shouldn't be in nixpkgs than anything else.

Furthermore, the complexity of this code is high. A once fairly simple module is now quite a bit more complex. We have to weigh the value of this feature against the maintenance cost it incurs upon anyone who works on the `auto-upgrade` mechanism in the future. There is significant cognitive load involved to understand this module with these changes. This is a high cost.

The value we get for that cost? Well, it's hard to say for sure, but this seems very much like a niche need to me. Which makes me wish it lived out of tree rather than in nixpkgs. When something is high in cost and low in impact, we've historically been fairly willing to remove it. This could live out of tree as an alternative implementation of the `auto-upgrade` mechanism.

Ekleog commented:
> The commit history in this PR needs some work.

I disagree: rebasing now would make life pointlessly hard for people who have already reviewed and only want to check the diff since their review. I'll just squash-and-merge when landing. See the rustc project guidelines for details

> The evaluation of a NixOS config is generally considered to be stateless, with the primary exception being the channels managed directly by the user / nix-channel. 

That's why the desyncs live in the `autoUpgrade` module, that is mainly designed to deal with channels and similar impurity. Desyncs don't add any more impurity than was already present in the autoUpgrade module: the desync state is recorded in automatically-managed files, just like the channel version is recorded in automatically-managed files.

> If the point is that there are packages that break so often that you'd like to have a desync ready for them perpetually, I think that's more of a sign that the package is not well-maintained and shouldn't be in nixpkgs than anything else.

That's exactly the point. I've had this exact issue with matrix-synapse, home-assistant and cups (due to proprietary drivers). I don't think we'd want to remove these 3 from nixpkgs.

> Furthermore, the complexity of this code is high.

That's definitely true, but it's also something pretty fundamental in how nixos works. If we were speaking of a module, or something less than 1% of the users would have enabled, I'd be aligned with you. But here we're speaking of the auto-uograde module, that's probably enabled on as many deployments as it's disabled on. I don't think having an alternative implementation of it would be a reasonable option, if only for discoverability reasons.

Ekleog commented on nixos/doc/manual/release-notes/rl-2505.section.md:
```diff
@@ -6,6 +6,10 @@
 
 Alongside many enhancements to NixOS modules and general system improvements, this release features the following highlights:
 
+- A `desync` option has been added to the `system.autoUpgrade` module, which allows upgrading the system's channels even when not all installed packages can be built. This works by continuing to track the problematic channel's previous revision, which is used for packages that are broken on the newer revision. Having multiple revisions of a channel can potentially lead to other issues, however ma ...
```
Uhh when I merged the 2511 file didn't exist yet, was it created after my merge commit? I wanted to use it too, but couldn't when I merged yesterday.

Anyway, with the delays this'll certainly still take, I'll adjust that on the next merge of main before landing, as 2511 will certainly exist by then.

Ekleog replied:
Nevermind, it's probably just my editor's file cache that hadn't updated or something like that — I've just moved the release notes to 25.11, thank you!

Ekleog committed:
no double-escape

Ekleog committed:
add links to docs

Ekleog committed:
fix release notes

Ekleog committed:
use pkgs.path

Ekleog committed:
fix mistake in merge resolution

Ekleog committed:
fix manual build

Ekleog committed:
third attempt at proper manual building, if manual took less time to build on my machine than on CI it'd be so great

Ekleog committed:
unlink a non-option to fix manual build

Ekleog committed:
hopefully fix ci

K900 commented:
Yeah, I don't think this is the way to go. Like, at all. What's the absolute smallest API surface we can expose in nixpkgs to allow for the rest of this to exist out of tree?

ElvishJerricco reviewed (changes requested):
It is a misuse of the commit power to dismiss someone else's review without their input. I will formulate a more in depth response later.

Ekleog commented:
@ElvishJerricco I'm not sure I understand why you're reacting so strongly. I guess you're annoyed at the stage this reached, but even if you saw this just now, it's been open for now 4 months and I literally pinged on at least 5 different days the nixos matrix channel. I'm not sure what I could have done more to help you notice it. To be honest I could have reasonably landed this like a month ago, when there were no strong objections against it and it had already been opened for 3 months. This would not have been a misuse of the commit bit.

I dismissed your review because I addressed all the points I could address, and I even requested another review from you. This has absolutely no relationship with commit power — if I had landed despite ongoing strong substantiated objections it would definitely have been, but I did not. I don't even think someone needs the commit bit to dismiss a review on their PR? I'm pretty sure I did that in the past on repositories I don't have commit rights to.

-----

@K900 I can't see any more minimal surface to implement this. The only alternative I can think of would be to only expose an `autoDesync` flag on containers, without exposing the desync infrastructure. But I'm relatively sure people would want to be able to desync outside of containers anyway, even if it goes beyond of the "stability guarantees of nixpkgs" that AFAICT are not actually guaranteed. And such a change would not actually lessen implementation complexity, just reduce the API surface — we still need the full implementation in order to get the benefits.

I hope my tone is not too bad in my answer, the recent exchanges have gotten me in a negative state of mind and I couldn't find any better words… and I can't answer other messages without answering yours too. Please forgive me if I'm sounding dismissive of your idea. This being said, I'm open to any idea you could have that'd allow to get the same behavior with less engineering on the nixpkgs side!

K900 commented:
I'm going to be honest here, I don't want this in nixpkgs. The implementation looks _okay_, but I really don't think this is something we should be shipping _no matter the implementation quality_. It's giving people a nudge in a very wrong direction, and it's going to massively complicate user support going forward. I see the value of this for people that can handle it with due care, but those people are in the extreme minority of our users.

ElvishJerricco reviewed (dismissed):
> I guess you're annoyed at the stage this reached, but even if you saw this just now, it's been open for now 4 months and I literally pinged on at least 5 different days the nixos matrix channel. I'm not sure what I could have done more to help you notice it.

It's my fault for not seeing it before. I am not annoyed that the PR has reached a late stage. I am concerned that we will merge something I think shouldn't be merged. I am truly sorry that you have spent a large amount of time working on this PR when I should have commented much sooner. That's my bad, but we can't fall into the sunk cost fallacy here. That it has received much work does not necessitate its merge.

---

> > The commit history in this PR needs some work.
>
> I disagree: rebasing now would make life pointlessly hard for people who have already reviewed and only want to check the diff since their review. I'll just squash-and-merge when landing. See the rustc project guidelines for details on the rationale.

This is not the rustc project. Now, to be fair, I don't think nixpkgs has a policy on how to merge, and we don't need to get into the arguments of squashes vs merges. But it is an overwhelming convention in nixpkgs that merge commits are used, retaining the merged branch's history. When rebasing on master or unstable, there shouldn't be any noteworthy changes to review. Or rather, if there are, they'll be just as hidden by merging in master as they would be by rebasing.

Furthermore, there are parts of this PR that probably *shouldn't* be one commit. Such as factoring out the `overlayType` type (in general, `/lib` changes should be self-contained commits). I think it's best to create the smallest form of the feature in one commit and add features to it in more commits. e.g. The `overlays` feature could be a separate commit.

If it's decided that this PR will be one big squash commit, then I guess these points will be moot. But I don't think it should be that way and I think most other nixpkgs committers agree with me. And in that case the branch history needs to be clean.

---

> Desyncs don't add any more impurity than was already present in the autoUpgrade module: the desync state is recorded in automatically-managed files, just like the channel version is recorded in automatically-managed files.

It is still *more* state and *more* impurity. Being of the same kind doesn't make it zero. And it's not exactly of the same kind anyway. These are not listed in `nix-channel --list`, for instance. You don't manage them with the same mechanism or knowledge.

> But here we're speaking of the auto-uograde module, that's probably enabled on as many deployments as it's disabled on.

Even if that's correct, which I doubt, I think the desync concept would be a small niche within that subset.

> > The first time a package breaks, you have to rollback your channel, add the desync, and then attempt the upgrade again.
>
> One thought I had was to introduce a script that'd automatically walk back the previous channel versions until it finds one that builds fine, and would use it. It's the way I've been doing manually, but I ended up deciding against it, to avoid even more complexity. Would such a tool make this PR more palatable to you?

Well, no, and that's part of the point. The machinery required to solve that problem is not worth having the problem solved IMO.

---

All in all, I think we have a couple fundamental issues to decide on somehow before this can be allowed to merge.

1. The squash merge thing.
2. Should this be in nixpkgs at all?

These questions can't be answered by editing code. We'll need to come to a consensus before this can be merged.

Ekleog commented on nixos/modules/tasks/auto-upgrade.nix:
```diff
@@ -178,12 +261,264 @@ in
+          Finally, note that overlays are not taken into account inside desyncs. This is
+          because, otherwise, evaluation would enter infinite loops.
+          If you need to overlay your desyncs, you can use the
+          [`desync.*.overlays`](#opt-system.autoUpgrade.desync._name_.overlays) option.
```
It should maybe have been changed to both at the same time. If this ends up being the only remaining hurdle, I'll be happy to add `{option}` before all options, but we do need the brackets to make the linking happen in practice.

winterqt reviewed (changes requested):
While I think this idea has merit, I agree with @ElvishJerricco and @K900 that this probably shouldn’t be in-tree.

> The added user support issues. To be honest, I don't think this is a problem. The description warnings are already largely dissuasive enough that people asking for support with a desync could just not see an answer and that'd be fine.

This alone is a reason we shouldn’t have this in-tree — people will use it in crazy ways, things *will* break, and the people who handle user support will regret that this was ever merged. I do not think descriptions are enough; we should just not be supplying users with footguns *at all*, no matter how useful.

Of course, nothing is stopping people from pulling this into their own configs — that’s the beauty of NixOS, after all — I just don’t think we should be adding new footguns, no matter how many past options we now wish were never added.

zimbatm commented:
@winterqt is your opinion that strong that it's worth losing @Ekleog as a contributor? I'm asking because sometimes we tend to get lost in the arguments and miss other aspects.

I can't speak for Ekleog of course but if I worked on a feature for 4 months, painstakingly working trough all the reviews and nitpicks, to then be blocked at the last minute, not based on strong technical argument that would be revealed, but based on hypothetical scenarios. I would be thinking twice before contributing anything meaninful to nixpkgs again. From an engineer's perspective, where I get reward from creating things, the only thing that's worse than that is getting my PR reverted.

I'm not saying there is no merit to your opinion, but we can't accurately predict the future either. A better way to find out, is to merge the PR, wait for the future to happen, and then change/remove the feature accordingly. Then ask @Ekleog if they are willing to help fix issues. That would feel more collaborative and construcitve to me and more in line with our values.

Ekleog commented:
@zimbatm Just to be transparent, this is not the only thing that's pushing me away in practice, and @winterqt actually answered after I had passed the turning point — I've had other PRs where contributing has been less than pleasant (the ongoing RFC, the rustic module, and maybe others I don't remember). I'm mostly of an overall feeling that the merging process has been made much more taxing, be it due to CI or human reviewers who are maybe just so exhausted they don't care about more than saying "no" any longer.

As you mention a PR being reverted it's actually "fun", because my first bad experience since coming back was actually my rustic PR getting reverted. I don't have issues with the revert itself: I do believe the committer who landed it was a bit too eager to land, and would have waited 1-2 weeks before doing so myself. But I do have issues with the way it was done, and the tone of exchanges.

All that to say, this PR is not a turning point for me, just the straw that broke the camel's back! I'll probably go back to only contributing small changes and just merging them myself for a few years, and come back then to see if I'm getting motivation again for bigger changes.

This being said, I do think that comments like "the whole feature is bad" should have come a few months ago when I first opened the PR and repeatedly pinged matrix, and not now that it was basically ready to land. @SuperSandro2000 had voiced that idea, but I was considering it handled by the various warnings. I think this is emblematic of the architectural issue @Ma27 mentioned, and I do believe that my several pings should have been enough considering the current governance.

-----

@Ma27 
> The big question si however, on what scale was this tested? How many different setups were involved? Are the scenarios theoretical or unrealistic?

I have tested it on 3 machines, all on nixos-unstable:
- one recent x86-64 laptop, only for cups inside a nixos container
- one old x86-64 server, for matrix-synapse, that I put in an overlay and outside any container — this is the "worst case scenario" I've tested I'm speaking of
- one aarch64 rpi, for home-assistant outside any container plus cups in a container

Cups inside nixos-container seems pretty wobbly, but I didn't hit any issue related to desyncs, just networking issues and similar.

These three machines have been happily living since I added them to desyncs, not blocking my updates. I haven't set metrics to check how much they were desynced in practice. But for both home-assistant and cups, I set them as desyncs at a time they were not building, and they worked fine as soon as I completed the process of setting up the desync.

IIRC both times there was already a PR up for the fix, but waiting for it would have meant a few more days of alerts and missed security fixes.

> Sorry for being pedantic here, but calling something "hypothetical" is was something I heard too often before seeing it blow up in production a few weeks later.

I totally understand this concern, and actually now would be the best time to land this — we've just released 25.05, which means we'd have a full 6 months of baking this in nixos-unstable before it reaches stable; and we could easily revert it until then, even though reverting after it reaches stable would be harder.

If we have stronger doubts, we could even rename it to autoUpgrade.experimental.desync.*; which would make it clear that it's a particularly experimental feature, even for nixos-unstable.

> Related, personally I don't want to end up in a situation where I don't voice my concerns about patches because I have to be afraid of being deemed responsible for making them quit.

I totally agree, and I'd like to be clear that I'm not threatquitting! I actually will most likely be stepping back regardless of the result here until I find some motivation again.

This being said, I do think there are solutions to the concerns voiced, like the "experimental" in the option name, that would allow to validate whether the concerns are valid or not; and the current answers I've been reading of "we should not do that regardless of any change" are IMO not good community building. (Just in case, I don't read your comment like this, and the questions you asked are definitely good questions!)

And to be fully transparent... the only good concern I've read why we should not do that regardless of any change was actually well-voiced by @roberth in their change of opinion — the risk of package support deteriorating is a real one that we can't really test better. This being said, I do strongly hope that any package maintainer would not desync their own packages, so I do hope we can try desyncs despite that drawback.

alyssais commented:
> We won't see any difference in quality of the packages because people still have to learn to patch them after 2 weeks.

This makes it more likely the fix will end up being done by one of the small number of extremely active, experienced Nixpkgs contributors for whom it's easier to fix the problem than ignore it, before those two weeks, rather than a new contributor for whom not having an easy way out might provide the boost they need to get over hesitance to contribute to Nixpkgs for the first time.

> re-assess at the next NixOS release how things are going.

How will we be able to evaluate that? There will be no way to tell how many contributions we would have received without this feature but haven't because of it.

zimbatm commented:
We don't have the infrastructure to measure it. That would be the same if desync was hosted somewhere else really. Last month we received contributions from 872 different people. Assuming a 1/100 contributor ratio, we have 89k users? If 100 of them adopt desync next month, we will lose 1 PR? Assuming the service is one maintained by the core contributors?

What we can see is how many issues are being raised against desync and check the user footgun claim, and gain more practical experience with it. Maybe we find out that desync kinda sucks, and in the process discover a better way to do things. That's what feature flags are for.

Even as an experienced nixpkgs contributor, sometimes it's nice to be able to apply a security upgrade immediately, and handle the side-quest afterwards. Especially in institutions where they have strict rules, I can see how a feature like this could encourage adoption, and lead to more contributors in aggregate.

I think the main concern of this PR is that it puts more pressure on core contributors. But I really think the risk is inflated. There are other bigger levers to pull. Like improving the tooling, process and contributor experience, that have a bigger impact than anything this PR could do. And if you asked @Ekleog nicely, maybe he could be enclined to help out.

Ekleog commented:
Thank you for the proposal! TBH I'm relatively happy with my current setup that just overrides the auto-upgrade module. If it could help other people if I upstream it to srvos then why not, but TBH I don't think I have the time/energy to do that without someone I'd know it'd help.

But thanks! I'll leave the PR open for the day we get proper governance, to decide whether we should close or land it :smile:

SuperSandro2000 commented:
> what's the point of NixOS compared to eg. Ansible on Fedora, if not the ability to mix packages at different versions?

The NixOS configuration is predictable and (ideally) fully declarative. Ansible depends on the state of the system before running it and the before state completely influences the outcome state. Running the same Ansible playbook on totally different hosts will result in totally different outputs system wise. Applying the same NixOS configuration to different systems ideally results in identically systems. Mixing different versions on NixOS is always a concise choice and not the outcome of mixing state accumulated over time.

And one other big point is, that you get a module which is known to be working with the shipped package without having to adjust multiple config files all over the place and having to configure all kinds of things over and over for every host and every user out there. You basically shift from "I build up a service and all its direct dependencies" to "I want to consume a service".

wolfgangwalther commented:
I looked at what the code does a bit more now and have a question for my understanding: Shouldn't this kind of thing be easily possible with your favorite dependency manager (flakes, npins, whatever) and some update tooling (for example https://github.com/Mic92/update-flake-inputs/)?

You can then just have two different references to nixpkgs and update them separately. Pull the desynced packages from one instance, the remaining system from the other. You can advance the latter more quickly than the former - auto-updates, built in CI and automatically merged, will just do that for you.

What am I missing that requires this to be a built-in feature in NixOS?

alyssais commented:
Perhaps we need a note of the risk in the documentation for the `system.autoUpgrade.enable`, as well as a recommendation to have some sort of monitoring for failed auto-upgrades?
